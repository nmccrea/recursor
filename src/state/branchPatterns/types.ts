import { EntityId } from "@reduxjs/toolkit"

export type Translation = Nominal<number, "Translation">
export type Scale = Nominal<number, "Scale">
export type Angle = Nominal<number, "Angle">
export type Depth = Nominal<number, "Depth">
export type Color = Nominal<string, "Color">

export interface BranchPatternNumericInputs {
  translation: Translation
  scale: Scale
  angle: Angle
  depth: Depth
}
export type BranchPatternNumericInputKey = keyof BranchPatternNumericInputs

export interface BranchPatternInputs extends BranchPatternNumericInputs {
  color: Color
}
export type BranchPatternInputKey = keyof BranchPatternInputs

export type BranchPatternId = Nominal<EntityId, "BranchPatternId">

export interface BranchPattern extends BranchPatternInputs {
  id: BranchPatternId
}
