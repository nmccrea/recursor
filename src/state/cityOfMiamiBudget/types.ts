export enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}

export interface CityOfMiamiBudgetState {
  status: Status
  data?: object
}
