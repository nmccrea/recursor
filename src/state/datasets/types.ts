export type Dataset = object

export type DatasetId = string

export interface DatasetIndex {
  [dataset: string]: DatasetState
}

export interface Datasets {
  index: DatasetIndex
}

export type DatasetState = {
  status: Status
  data?: Dataset
}

export interface FetchResult {
  datasetId: DatasetId
  data: Dataset
}

export enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}
