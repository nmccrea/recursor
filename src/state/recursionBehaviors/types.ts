export type Translation = Nominal<number, "Translation">
export type Scale = Nominal<number, "Scale">
export type Angle = Nominal<number, "Angle">
export type Depth = Nominal<number, "Depth">
export type Color = Nominal<string, "Color">

export interface Transformation {
  translation: Translation
  scale: Scale
  angle: Angle
}

export type RecursionBehaviorId = Nominal<string, "RecursionBehaviorId">

export interface RecursionBehavior extends Transformation {
  id: RecursionBehaviorId
  depth: Depth
  color: Color
}
