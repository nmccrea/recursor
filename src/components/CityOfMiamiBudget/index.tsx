/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import React from "react"
import { useDispatch, useSelector } from "react-redux"
import { fetchDataset } from "../../state/datasets/async"
import {
  selectorForAsyncState,
  selectorForData,
} from "../../state/datasets/selectors"
import { DatasetId, AsyncState } from "../../state/datasets/types"

const DATASET_ID: DatasetId = "cityOfMiami/budget/revenue"
const DATA_URL = "https://data.miamigov.com/resource/ub3m-qgg5.json"

const Data = () => {
  const data = useSelector(selectorForData(DATASET_ID))
  return (
    <>
      <h3>The Data:</h3>
      <p>{JSON.stringify(data)}</p>
    </>
  )
}

const ELEMENT_FOR_ASYNC_STATE = {
  [AsyncState.Idle]: null,
  [AsyncState.Pending]: <p>Loading...</p>,
  [AsyncState.Fulfilled]: <Data />,
}

const CityOfMiamiBudget = () => {
  const asyncState = useSelector(selectorForAsyncState(DATASET_ID))
  if (asyncState === undefined || asyncState === AsyncState.Idle)
    useDispatch()(fetchDataset(DATA_URL, DATASET_ID))
  return (asyncState && ELEMENT_FOR_ASYNC_STATE[asyncState]) || null
}

export default CityOfMiamiBudget
