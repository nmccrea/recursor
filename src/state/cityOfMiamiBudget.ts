/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import { RootState } from "./store"
import {
  createAsyncThunk,
  createSlice,
  CaseReducer,
  PayloadAction,
} from "@reduxjs/toolkit"

// State
enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}

interface CityOfMiamiBudgetState {
  status: Status
  data?: object
}

const INITIAL_STATE: CityOfMiamiBudgetState = { status: Status.Idle }

// Case Reducers
const fetchDataPending: CaseReducer<CityOfMiamiBudgetState, PayloadAction> = (
  previousState
) => ({
  ...previousState,
  status: Status.Pending,
})

const fetchDataFulfilled: CaseReducer<
  CityOfMiamiBudgetState,
  PayloadAction<object>
> = (previousState, action) => ({
  ...previousState,
  status: Status.Fulfilled,
  data: action.payload,
})

// Async Thunks
const fetchData = createAsyncThunk("fetchData", async (url: string) => {
  const response = await fetch(url)
  return await response.json()
})

// Slice
const { reducer } = createSlice({
  name: "cityOfMiamiBudget",
  initialState: INITIAL_STATE,
  reducers: {},
  extraReducers: {
    [fetchData.pending.type]: fetchDataPending,
    [fetchData.fulfilled.type]: fetchDataFulfilled,
  },
})

// Selectors
const getStatus = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.status) || undefined

const getData = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.data) || undefined

export default reducer
export { fetchData, getStatus, getData, Status }
