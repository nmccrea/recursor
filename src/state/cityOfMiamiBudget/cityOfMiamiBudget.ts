/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import { createSlice } from "@reduxjs/toolkit"
import { fetchData, fetchDataPending, fetchDataFulfilled } from "./async"

// State
export enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}

export interface CityOfMiamiBudgetState {
  status: Status
  data?: object
}

const INITIAL_STATE: CityOfMiamiBudgetState = { status: Status.Idle }

// Slice
const { reducer } = createSlice({
  name: "cityOfMiamiBudget",
  initialState: INITIAL_STATE,
  reducers: {},
  extraReducers: {
    [fetchData.pending.type]: fetchDataPending,
    [fetchData.fulfilled.type]: fetchDataFulfilled,
  },
})

export default reducer
