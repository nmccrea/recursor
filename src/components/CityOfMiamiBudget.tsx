/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import React from "react"
import { useDispatch, useSelector } from "react-redux"
import { fetchData } from "../state/cityOfMiamiBudget/async"
import { Status } from "../state/cityOfMiamiBudget/types"
import { getStatus, getData } from "../state/cityOfMiamiBudget/selectors"

const DATA_URL = "https://data.miamigov.com/resource/ub3m-qgg5.json"

const FetchButton = () => {
  const dispatch = useDispatch()
  return (
    <button onClick={() => dispatch(fetchData(DATA_URL))}>Fetch Data!</button>
  )
}

const Loading = () => <p>Loading...</p>

const Data = () => {
  const data = useSelector(getData)
  return (
    <>
      <h3>The Data:</h3>
      <p>{JSON.stringify(data)}</p>
    </>
  )
}

const ELEMENT_FOR_STATUS = {
  [Status.Idle]: <FetchButton />,
  [Status.Pending]: <Loading />,
  [Status.Fulfilled]: <Data />,
}

const CityOfMiamiBudget = () => {
  const status = useSelector(getStatus)
  return ELEMENT_FOR_STATUS[status]
}

export default CityOfMiamiBudget
