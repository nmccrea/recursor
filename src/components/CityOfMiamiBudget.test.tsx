import React from "react"
import * as reactRedux from "react-redux"
import { render } from "@testing-library/react"
import { Status } from "../state/cityOfMiamiBudget/types"
import CityOfMiamiBudget from "./CityOfMiamiBudget"

describe("<CityOfMiamiBudget />", () => {
  it("renders correctly when status is idle", () => {
    jest.spyOn(reactRedux, "useSelector").mockReturnValue(Status.Idle)
    jest.spyOn(reactRedux, "useDispatch").mockReturnValue(jest.fn())

    const { container } = render(<CityOfMiamiBudget />)

    expect(container).toMatchSnapshot()
  })

  it("renders correctly when status is pending", () => {
    jest.spyOn(reactRedux, "useSelector").mockReturnValue(Status.Pending)
    jest.spyOn(reactRedux, "useDispatch").mockReturnValue(jest.fn())

    const { container } = render(<CityOfMiamiBudget />)

    expect(container).toMatchSnapshot()
  })

  it("renders correctly when status is fulfilled", () => {
    jest
      .spyOn(reactRedux, "useSelector")
      .mockReturnValueOnce(Status.Fulfilled)
      .mockReturnValueOnce({ a: "apples", b: "bananas" })
    jest.spyOn(reactRedux, "useDispatch").mockReturnValue(jest.fn())

    const { container } = render(<CityOfMiamiBudget />)

    expect(container).toMatchSnapshot()
  })
})
