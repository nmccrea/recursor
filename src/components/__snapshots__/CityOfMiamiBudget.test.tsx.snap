// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CityOfMiamiBudget /> renders correctly when status is fulfilled 1`] = `
<div>
  <h3>
    The Data:
  </h3>
  <p>
    {"a":"apples","b":"bananas"}
  </p>
</div>
`;

exports[`<CityOfMiamiBudget /> renders correctly when status is idle 1`] = `
<div>
  <button>
    Fetch Data!
  </button>
</div>
`;

exports[`<CityOfMiamiBudget /> renders correctly when status is pending 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;
